http/basic/src/main/java/io/cloudevents/http/HttpMessageFactory.java

```diff
@@ -0,0 +1,166 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cloudevents.http;
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpsExchange;
+import io.cloudevents.core.message.MessageReader;
+import io.cloudevents.core.message.MessageWriter;
+import io.cloudevents.core.message.impl.GenericStructuredMessageReader;
+import io.cloudevents.core.message.impl.MessageUtils;
+import io.cloudevents.core.message.impl.UnknownEncodingMessageReader;
+import io.cloudevents.http.impl.CloudEventsHeaders;
+import io.cloudevents.http.impl.HttpMessageReader;
+import io.cloudevents.http.impl.HttpMessageWriter;
+
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URLConnection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import static io.cloudevents.http.impl.CloudEventsHeaders.CONTENT_TYPE;
+
+/**
+ * This class provides a collection of methods to create {@link io.cloudevents.core.message.MessageReader}
+ * and {@link io.cloudevents.core.message.MessageWriter} for various HTTP APIs.
+ */
+public final class HttpMessageFactory {
+
+    private HttpMessageFactory() {}
+
+    /**
+     * Creates a new {@link MessageReader} that can read both structured and binary messages from a HTTP response (client) or request (server).
+     *
+     * <pre>
+     * Example of usage with <a href="https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html">HttpServletRequest</a>:
+     * {@code
+     * Consumer<BiConsumer<String,String>> forEachHeader = processHeader -> {
+     *     Enumeration<String> headerNames = httpServletRequest.getHeaderNames();
+     *     while (headerNames.hasMoreElements()) {
+     *         String name = headerNames.nextElement();
+     *         processHeader.accept(name, httpServletRequest.getHeader(name));
+     *
+     *     }
+     * };
+     * byte[] body;
+     * String contentLength = httpServletRequest.getHeader("Content-Length");
+     * if (contentLength != null) {
+     *     body = IOUtils.toByteArray(httpServletRequest.getInputStream(), Integer.parseInt(contentLength));
+     * } else {
+     *     body = IOUtils.toByteArray(httpServletRequest.getInputStream());
+     * }
+     * HttpMessageFactory.createReader(forEachHeader, body);
+     * }
+     * </pre>
+     * @param forEachHeader http headers visitor function
+     * @param body          nullable buffer of the body
+     * @return a message reader implementation with potentially an unknown encoding
+     * @throws IllegalArgumentException If, in case of binary mode, the spec version is invalid
+     */
+    public static MessageReader createReader(Consumer<BiConsumer<String, String>> forEachHeader, byte[] body) {
+        final String[] hdrs = {null, null};
```

Comment:
There's no really good solution here, but I think it would be better to have
```
AtomicReference<String> contentType = new AtomicReference<>();
AtomicReference<String> specVersion = new AtomicReference<>();
```
and then use `contentType.get()` and `specVersion.get()` directly in the call to `parseStructuredOrBinaryMessage`.

Reply:
I was thinking of of create private static class with the two properties.

Reply:
Yep +1 for `AtomicReference`, the arrays are kinda weird. I think a class here is overkilling too